### PR TITLE
Feat: Add automation script to generate component

### PR DESCRIPTION
**Pull request by ArthurTriis1**
## What's the purpose of this pull request?
We noticed that every time we wanted to create a new component, we created many new files. So, we thought it might be a good idea to generate these default templates automatically. It will help our development experience and we can waste less time making the "boilerplates files"  and jump into the component itself, decreasing the chances of forgetting to update/create a file.

We are using [Plop](https://plopjs.com) framework to generate those files.

## How it works? 

- run `yarn generate`

```bash
# What is your component name? 

ComponentName

#  What is your component atomic group?  (Use arrow keys)

> atoms 
> molecules 
> organisms
```

- it will create all the base files with the component given name;
- It will create a folder for the component under its atomic group;

- Files generated automatically:
```
✔  ++ faststore/packages/ui/src/atoms/ComponentName/ComponentName.tsx
✔  ++ faststore/packages/ui/src/atoms/ComponentName/index.tsx
✔  ++ faststore/themes/theme-b2c-tailwind/src/atoms/component-name.css
✔  ++ faststore/packages/ui/src/atoms/ComponentName/ComponentName.test.tsx
✔  ++ faststore/packages/ui/src/atoms/ComponentName/stories/ComponentName.stories.tsx
✔  ++ faststore/packages/ui/src/atoms/ComponentName/stories/ComponentName.mdx
✔  _+ faststore/packages/ui/src/index.ts
```

And that is it! You will be ready to work on your component. 🎉
* For the component name, please follow CamelCase pattern, and do not use `-`, i.e, `Component-Name`
* Please try to not rename `// Atoms`, `// Molecules`, `// Organisms` on `packages/ui/scr/index.ts` [file](https://github.com/vtex/faststore/blob/b1badead35dc43b41f25480a66f4e81cc9a8e22a/packages/ui/src/index.ts), because we are using it as [pattern](https://github.com/vtex/faststore/blob/443cbd117e34ea1702a2994ca4ad0b2733f17189/generators/plopfile.ts#L62) to update it.

We are using templates files just with the base structure according to what we find helpful to have. You can take a look into it on `generators/templates` folder.

Are you missing any files? Or is there anything else we can generate automatically? Any suggestions? Please, feel free to leave a comment below 🙃

## References
- [Plop js](https://plopjs.com): "micro-generator framework"

### CODE REVIEW

**Comment by hellofanny** [comment]
Is there any readme file that we can add this information? 👀

**Comment by tlgimenes** [comment]
> Is there any readme file that we can add this information? 👀

Maybe add it on contributing.md?

**Review by tlgimenes — APPROVED** [review]
AMAZING WORK! 🎉🎉🎉

**Comment by igorbrasileiro** [comment]
😲 Good addition @ArthurTriis1 !! Can you add a brief instruction inside the **docs/component-architecture.md** ?

**Review by igorbrasileiro — APPROVED** [review]

**Review by lucasfp13 — APPROVED** [review]
Great job, guys! 🎉

File: CONTRIBUTING.MD [hunk 1]
```diff
@@ -34,6 +34,42 @@ Example:
...
+✔  ++ faststore/packages/ui/src/atoms/ComponentName/stories/ComponentName.stories.tsx
+✔  ++ faststore/packages/ui/src/atoms/ComponentName/stories/ComponentName.mdx
+✔  _+ faststore/packages/ui/src/index.ts
+```
+
+And that is it! You will be ready to work on your component. 🎉
+
+> For the component name, please follow CamelCase pattern, and do not use `-`, i.e., Component-Name.
+
+> Please try to not edit // Atoms, // Molecules, // Organisms comments on `packages/ui/scr/index.ts` [file](https://github.com/vtex/faststore/blob/b1badead35dc43b41f25480a66f4e81cc9a8e22a/packages/ui/src/index.ts), because we are using it as [pattern](https://github.com/vtex/faststore/blob/443cbd117e34ea1702a2994ca4ad0b2733f17189/generators/plopfile.ts#L62) to update it. For more details, check [here](https://github.com/vtex/faststore/pull/1039).
```
Comment (lariciamota) [inline CONTRIBUTING.MD:71]:
```suggestion
> Please do not edit `// Atoms`, `// Molecules` and `// Organisms` comments on `packages/ui/scr/index.ts` [file](https://github.com/vtex/faststore/blob/b1badead35dc43b41f25480a66f4e81cc9a8e22a/packages/ui/src/index.ts), because we are using it as a [pattern](https://github.com/vtex/faststore/blob/443cbd117e34ea1702a2994ca4ad0b2733f17189/generators/plopfile.ts#L62) to update the file. For more details, check [here](https://github.com/vtex/faststore/pull/1039).
```

File: CONTRIBUTING.MD [hunk 2]
```diff
@@ -34,6 +34,42 @@ Example:
...
+✔  ++ faststore/themes/theme-b2c-tailwind/src/atoms/component-name.css
+✔  ++ faststore/packages/ui/src/atoms/ComponentName/ComponentName.test.tsx
+✔  ++ faststore/packages/ui/src/atoms/ComponentName/stories/ComponentName.stories.tsx
+✔  ++ faststore/packages/ui/src/atoms/ComponentName/stories/ComponentName.mdx
+✔  _+ faststore/packages/ui/src/index.ts
+```
+
+And that is it! You will be ready to work on your component. 🎉
+
+> For the component name, please follow CamelCase pattern, and do not use `-`, i.e., Component-Name.
```
Comment (lariciamota) [inline CONTRIBUTING.MD:69]:
```suggestion
> For the component name, please follow the Pascal Case pattern, and do not use `-`, i.e., Component-Name.
```

File: generators/templates/Component.tsx.hbs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import React from 'react'
+
+export interface {{pascalCase name}}Props {
+  /**
+   * ID to find this component in testing tools (e.g.: cypress,
+   * testing-library, and jest).
+   */
+  testId?: string
+  children?: React.ReactNode
```
Comment (lariciamota) [inline generators/templates/Component.tsx.hbs:9]: Why is this line necessary? 
```suggestion
```

**Review by lariciamota — COMMENTED** [review]
Nice!! 🚀

File: package.json [hunk 4]
```diff
@@ -48,6 +49,7 @@
     "husky": "^4.3.0",
     "lerna": "^3.22.1",
     "lint-staged": "^10.5.1",
+    "plop": "^2.7.6",
```
Comment (filipewl) [inline package.json:52]: Hmm. I'm wondering if this package should be here, at the root of the monorepo, or at `packages/ui/` since it's related to that. 

Any thoughts? 🤔

Comment (filipewl) [reply package.json:52]: Oh, [I see now](https://github.com/vtex/faststore/pull/1039/files#diff-d389cc1269e8b4c6b14b577978948d0380a8437dacb734fc319a4a66729e238bR36) it needs access to a previous directory file such as `themes/theme-b2c-tailwind`.

**Comment by hellofanny** [comment]
> 😲 Good addition @ArthurTriis1 !! Can you add a brief instruction inside the **docs/component-architecture.md** ?

@ArthurTriis1 Actually, I think we can move the instruction from `CONTRIBUTING.MD` to `*docs/component-architecture.md`. It makes more sense. 🤔

File: generators/templates/Component.tsx.hbs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import React from 'react'
```
Comment (lariciamota) [inline generators/templates/Component.tsx.hbs:1]:
```suggestion
import React, { forwardRef }  from 'react'
```

File: generators/templates/Component.tsx.hbs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+export interface {{pascalCase name}}Props {
+  /**
+   * ID to find this component in testing tools (e.g.: cypress,
+   * testing-library, and jest).
+   */
+  testId?: string
+}
+
+const {{pascalCase name}} = ({ testId = 'store-{{kebabCase name}}', children }: {{pascalCase name}}Props) => {
```
Comment (lariciamota) [inline generators/templates/Component.tsx.hbs:11]: I commented about children, but I didn't comment on this line also. In order for it to work the prop interface would need to export something that has children on it or we could use the `forwardRef` here, I prefer the latter.
```suggestion
const {{pascalCase name}} = forwardRef<HTMLDivElement, {{pascalCase name}}Props>(function {{pascalCase name}}(
   { testId = 'store-{{kebabCase name}}', children, ...otherProps },
  ref
) {
```

File: generators/templates/Component.tsx.hbs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  return (
+    <div data-store-{{kebabCase name}} data-testid={testId}>
```
Comment (lariciamota) [inline generators/templates/Component.tsx.hbs:13]:
```suggestion
    <div ref={ref} data-store-{{kebabCase name}} data-testid={testId} {...otherProps}>
```

File: package.json [hunk 8]
```diff
@@ -17,7 +17,8 @@
     "test": "lerna run test --",
     "bootstrap": "lerna bootstrap",
     "release": "lerna version minor --yes && lerna publish from-git --yes",
-    "size": "lerna run size"
+    "size": "lerna run size",
+    "generate": "yarn plop --plopfile generators/plopfile.ts"
```
Comment (filipewl) [inline package.json:21]: Since the generator is at the monorepo's root, so it's not technically something just related to the UI project, I feel like we need to be more descriptive with the command's name. It's not immediately clear **what** is going to be generated when you run the command at the root. 

Just `generate` would be fine if it was inside `packages/ui`, but since it's here, I suggest we use something like `generate-ui-component`.

Or maybe I'm exaggerating.
😆

Comment (lariciamota) [reply package.json:21]: I liked this suggestion of being more specific about what is being generated!

**Review by saranicoly — APPROVED** [review]
Great job guys! It will be very useful.

**Comment by eduardoformiga** [comment]
Really good job here, @ArthurTriis1 ✨ 

Do we still need to work? hahaha
